src/models/pairwise_union_reach_surface.py

```diff
@@ -0,0 +1,136 @@
+      reach_points: A list of ReachPoints to which the model is to be fit. This
+        is parallel to the reach_curves list. The reach point at ith poisition
+        is drawn from ith reach curve.
+    """
+
+    self._reach_curves = copy.deepcopy(reach_curves)
+    self._n = len(reach_points)
+    self._p = len(reach_points[0].impressions)
+    super().__init__(data=reach_points, max_reach=0)
+
+  def _fit(self) -> None:
+    z, alpha = self.get_z_and_alpha()
+    self._a = self.solve_a_given_z_and_alpha(z, alpha)
+
+  def get_reach_vector(self, impressions: Iterable[int]) -> Iterable[int]:
+    return [
+        reach_curve.by_impressions(impression).reach()
+        for reach_curve, impression in zip(self._reach_curves, impressions)
+    ]
+
+  def by_impressions(self,
+                     impressions: Iterable[int],
+                     max_frequency: int = 1) -> ReachPoint:
+
+    reach_vector = self.get_reach_vector(impressions)
+    overlapped_reach = sum(reach_vector)
+    overlap = sum([
+        (self._a[i * self._p + j] * reach_vector[i] * reach_vector[j]) /
+        (max(self._reach_curves[i].max_reach, self._reach_curves[j].max_reach) *
+         2) for i in range(self._p) for j in range(self._p)
+    ])
+
+    return ReachPoint(impressions, [overlapped_reach - overlap])
+
+  def get_constraints(self):
+    G1 = -matrix(np.eye(self._p * self._p))
+    G2 = np.kron(np.eye(self._p), np.ones(self._p))
+    G3 = np.kron(np.ones(self._p), np.eye(self._p))
+    h = matrix(
+        np.concatenate((np.zeros(self._p * self._p), np.ones(2 * self._p)),
+                       axis=None))
+    return matrix(np.vstack((G1, G2, G3))), h
+
+  def construct_matrix_a(self):
+    A = np.zeros(shape=(self._p, self._p * self._p))
+    for i in range(self._p):
+      A[i, i * self._p + i] = 1
+    return matrix(A)
+
+  def solve_a_given_z_and_alpha(self, z, alpha):
+    """Optimize for matrix a given z and alpha.
+
+    Args:
+      z: a length-n list of length-(p * p) vectors.
+      alpha: a length-n list of scalars.
+
+    Returns:
+      a length-(p * p) vector of the optimization problem above
+    """
+
+    P = matrix(
+        sum([
+            np.matmul(z[k].reshape(self._p * self._p, 1),
+                      z[k].reshape(1, self._p * self._p))
+            for k in range(self._n)
+        ]))
+    q = -matrix(
+        sum((self._data[k].reach() - alpha[k]) * z[k] for k in range(self._n)))
+
+    G, h = self.get_constraints()
+    A = self.construct_matrix_a()
+    b = matrix(np.zeros(self._p))
+    res = solvers.qp(P, q, G, h, A, b)
+
+    return np.array(res['x']).flatten()
+
+  def get_z_and_alpha(self):
+    """Get intermidiate quantities z and alpha.
+
+    Returns:
+      Tuple of z and alpha.
+      z: Length-n list of length-(p * p) vectors. Each vector is flattened
+      version of the matrix that denotes pairwise overlap regions.
+      alpha: Length-n list of scalars. Each scalar denotes the overlapped reach
```

Comment:
Is alpha the same as b in the doc?

Reply:
no Matthew, z is equivalent to B in the doc.
and d in the doc is ->    d = r_hat(true reach) - alpha

does this make sense? I also outlined how the qp matrices are created in the docstring. If you prefer, I can definitely change the name of the variables.